tools/azsdk-cli/Azure.Sdk.Tools.Cli/Tools/Package/SampleGeneration/DotNetSourceInputProvider.cs

```diff
@@ -0,0 +1,167 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System.Collections.Concurrent;
+using Azure.Sdk.Tools.Cli.Helpers;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+
+namespace Azure.Sdk.Tools.Cli.SampleGeneration;
+
+/// <summary>
+/// Discovers C# source and supporting infrastructure files for .NET client libraries.
+/// </summary>
+public class DotNetSourceInputProvider : ILanguageSourceInputProvider
+{
+
+    private static readonly string[] s_testFileNameIndicators =
+    [
+        "Tests.cs",
+        "Test.cs",
+        "Facts.cs",
+        "Spec.cs",
+    ];
+
+    private static readonly HashSet<string> s_testMethodAttributes = new(StringComparer.OrdinalIgnoreCase)
+    { "Test", "Fact", "Theory", "TestCase", "DataTestMethod", "Repeat" };
+
+    private static readonly HashSet<string> s_lifecycleAttributes = new(StringComparer.OrdinalIgnoreCase)
+    { "SetUp", "TearDown", "OneTimeSetUp", "OneTimeTearDown", "ClassInitialize", "ClassCleanup", "AssemblyInitialize", "AssemblyCleanup" };
+
+    private static readonly HashSet<string> s_infraBaseTypeIndicators = new(StringComparer.OrdinalIgnoreCase)
+    { "RecordedTestBase", "ClientTestBase", "SamplesBase" };
+
+    private static readonly ConcurrentDictionary<string, (DateTime LastWrite, bool IsTest)> s_cache = new();
+
+    public IReadOnlyList<FileHelper.SourceInput> Create(string packagePath)
+    {
+        var result = new List<FileHelper.SourceInput>();
+
+        var src = Path.Combine(packagePath, "src");
+        if (Directory.Exists(src))
+        {
+            result.Add(new FileHelper.SourceInput(src, IncludeExtensions: [".cs"]));
+        }
+
+        var tests = Path.Combine(packagePath, "tests");
+        if (Directory.Exists(tests))
+        {
+            foreach (var file in EnumerateInfrastructureFiles(tests))
+            {
+                result.Add(new FileHelper.SourceInput(file));
+            }
+        }
+
+        var samples = Path.Combine(packagePath, "tests", "samples");
+        if (Directory.Exists(samples))
+        {
+            result.Add(new FileHelper.SourceInput(samples, IncludeExtensions: [".cs"]));
+        }
+
+        return result;
+    }
+
+    private static IEnumerable<string> EnumerateInfrastructureFiles(string testsDirectory)
```

Comment:
What is an "infrastructure file" in this context and why is it needed as input here?

Reply:
Files that abstract the values of the test resources.